### ISSUE TITLE
Accessing action arguments from store listeners possible?

### ISSUE THREAD

**Issue by aldanor** [goatslacker/alt#462]
Here's a simple example that I've been trying to solve: 

- Store's state is essentially one big dict
- There's a top-level container component whose state is bound to store's state
- There's an action `updateValue(key, value)` which updates a value for a single key and which may be called from child components. This may or may not change store's state (the store is an `Immutable.Map`)

Now the problem is, I have a socket connection which needs to send incremental updates every time an update happens, like `{key: 'foo', value: 'bar'}`, but only if the store's state actually changed. There's a few naive options I see:

- Subscribe the socket connection object directly to the store, but then it would only be able to access the entire state through `getState`, and it wouldn't know what exactly changed
- Send a socket update imperatively in the action function before dispatching -- this would couple actions to socket logic and it wouldn't be able to discern whether the store's state changed or not.

Is there a better way of doing this or am I missing something? (I've been able to do this in Reflux quite easily)

TL;DR if an action is fired with some arguments, can the store's listener somehow access those?

**Comment by goatslacker** [comment]
well yeah, here's a naive example which doesn't have the correct syntax but demonstrates how to

```js
actions.foo(1)

store.onFoo = function (one) {
  one === 1
}
```

**Comment by aldanor** [comment]
@goatslacker Thanks for the reply!

But that wasn't the question really :) Of course the *store* can access the action parameters.

But the store's *listeners* can't.

**Comment by goatslacker** [comment]
ah sorry, yes that's very intentional. If you'd like something like this you can setup a store that listens to every action and sets the last received value:

```js
const LastValueStore = alt.createStore({
  reduce(currentState, payload) {
    return payload.data
  }
})

LastValueStore.listen((lastValue) => {
  console.log(lastValue)
})
```

Alternatively, you can create a property in your state that holds this last value and you can query that.

**Comment by aldanor** [comment]
Yep, that's what I was thinking, exactly. But then, there are no guarantees that `LastValueStore` will actually receive all updates? It may receive *some* of them, right?

So it would then basically have to keep the entire update history...

**Comment by goatslacker** [comment]
It'll receive all updates sure. The event loop is synchronous so you'll get an event emitted every time an action is dispatched through the loop.

**Comment by aldanor** [comment]
@goatslacker So if more actions were fired while the store was processing the action, then when the store emits the change, all its listeners would first have to finish their work first before the store starts processing the next action?

If that's so, that's very cool and is actually much simpler than I thought :)

// it's not quite obvious from the docs though; maybe I'm missing something

**Comment by goatslacker** [comment]
if you fire off 4 actions the store will emit 4 times.

action a fires
store emits with info from a
action b fires
store emits with info from b
...

you won't get into a case where

action a fires
action b fires
store...

because it's all synchronous

**Comment by aldanor** [comment]
@goatslacker Does this also apply if the action was called as `action.defer(...)`?

**Comment by goatslacker** [comment]
Yes

**Comment by aldanor** [comment]
Gotcha, thanks!
